import { Repository } from "typeorm";
import bcrypt from "bcrypt";

import { User } from "../entity/User";
import { UserData } from "../types";
import createHttpError from "http-errors";
import { Roles } from "../constants";

export class UserService {
    constructor(private userRepository: Repository<User>) {}

    async create({ firstName, lastName, email, password }: UserData) {
        const user = await this.userRepository.findOne({
            where: { email: email },
        });

        if (user) {
            const err = createHttpError(400, "Email is already exists!");
            throw err;
        }

        // Hash the password
        const saltRounds = 10;
        const hashPassword = await bcrypt.hash(password, saltRounds);

        try {
            const user = await this.userRepository.save({
                firstName,
                lastName,
                email,
                password: hashPassword,
                role: Roles.CUSTOMER,
            });

            return user;
        } catch (err) {
            const error = createHttpError(
                500,
                "Failed to store data in the databse",
            );
            throw error;
        }
    }

    async findByEmail(email: string) {
        try {
            const user = await this.userRepository.findOne({
                where: { email },
            });

            return user;
        } catch (err) {
            if (err instanceof createHttpError.HttpError) {
                throw err;
            }

            const error = createHttpError(500, "Failed to find user by email");
            throw error;
        }
    }

    async findById(id: number) {
        return await this.userRepository.findOne({
            where: { id },
        });
    }
}
